common/subtitle-coloring/subtitle-coloring.ts

```diff
@@ -0,0 +1,790 @@
+const ANKI_RECENTLY_MODIFIED_INTERVAL = 10000;
+const HAS_LETTER_REGEX = /\p{L}/u;
+
+interface TrackState {
+    track: number;
+    dt: DictionaryTrack;
+    yomitanHealthy: boolean;
+    tokenizeCache: Map<string, string[]>;
+    tokenStatusCache: Map<string, TokenStatus | null>;
+    lemmatizeCache: Map<string, string[]>;
+    ankiCardIdStatuses: Map<number, TokenStatus>;
+    ankiSuspendedCardIds: Set<number>;
+}
+
+export class SubtitleColoring extends SubtitleCollection<RichSubtitleModel> {
+    private _subtitles: RichSubtitleModel[];
+    private settings?: AsbplayerSettings;
+    private subtitlesInterval?: NodeJS.Timeout;
+    private showingSubtitles?: RichSubtitleModel[];
+    private showingNeedsRefreshCount: number;
+
+    private yomitan: Yomitan;
+    private anki: Anki | undefined;
+    private readonly fetcher?: Fetcher;
+    private trackStates: TrackState[];
+    private erroredCache: Set<number>;
+    private uncollectedCache: Set<number>;
+    private uncollectedNeedsRefresh: boolean;
+    private ankiRecentlyModifiedCardIds: Set<number>;
+    private ankiLastRecentlyModifiedCheck: number;
+    private ankiRecentlyModifiedFirstCheck: boolean;
+    private colorCacheLastRefresh: number;
+    private colorCacheBuilding: boolean;
+    private colorCacheBuildingCurrentIndexes: Set<number>;
+    private shouldCancelBuild: boolean; // Set to true to stop current color cache build, checked after each async call
+    private tokenRequestFailed: boolean;
+
+    private readonly subtitleColorsUpdated: (updatedSubtitles: RichSubtitleModel[]) => void;
+    private readonly getMediaTimeMs?: () => number;
+
+    constructor(
+        settingsPromise: Promise<AsbplayerSettings>,
+        options: SubtitleCollectionOptions,
+        subtitleColorsUpdated: (updatedSubtitles: RichSubtitleModel[]) => void,
+        getMediaTimeMs?: () => number,
+        fetcher?: Fetcher
+    ) {
+        super([], { ...options, returnNextToShow: true });
+        this._subtitles = [];
+        settingsPromise.then((settings) => (this.settings = settings));
+        this.fetcher = fetcher;
+        this.yomitan = new Yomitan(this.fetcher);
+        this.trackStates = [];
+        this.subtitleColorsUpdated = subtitleColorsUpdated;
+        this.getMediaTimeMs = getMediaTimeMs;
+        this.showingNeedsRefreshCount = 0;
+        this.erroredCache = new Set();
+        this.uncollectedCache = new Set();
+        this.uncollectedNeedsRefresh = false;
+        this.ankiRecentlyModifiedCardIds = new Set();
+        this.ankiLastRecentlyModifiedCheck = Date.now();
+        this.ankiRecentlyModifiedFirstCheck = true;
+        this.colorCacheLastRefresh = Date.now();
+        this.colorCacheBuilding = false;
+        this.colorCacheBuildingCurrentIndexes = new Set();
+        this.shouldCancelBuild = false;
+        this.tokenRequestFailed = false;
+    }
+
+    get subtitles() {
+        return this._subtitles;
+    }
+
+    set subtitles(subtitles) {
```

Comment:
Since you're already extending `SubtitleCollection` I think code will get a little simpler if you override `initSubtitleCollection` here and just wrap a call to the `super` implementation with the same code. That will get rid of some `instanceof` checks above. You could consider renaming `initSubtitleCollection` to something more abstract if the current name feels too weird for that.

Alternatively, it seems like the only reason to have a setter like this is to share cached state. I think that would still be possible even without a mutator method like this.